### ISSUE TITLE
[FEATURE] Roadmap

### ISSUE THREAD

**Issue by Aeres-u99** [hwittenborn/makedeb#16]
From our chats: 

```
Hunter Wittenborn: 
> I would like to have a roadmap, basically the feature requests/debugging/analysis and anything 
of sorts and I would like it listed somewhere,
 this way we can just take up something and know that xyz is working on it and be aware of how things are going.
Definitely, I think that'd help a ton if anyone else was wanting to get involved.

Mind making a GitHub issue on the topic? 
```
This issue is just a gentle reminder for roadmap asap. Would facillitate better development and contribution

**Comment by hwittenborn** [comment]
Great! Glad to continue it all here.

Continuing from the support rooms then - is there any system you had in mind? The immediate thing that comes to mind is GitHub's [Projects](https://docs.github.com/en/issues/organizing-your-work-with-project-boards/managing-project-boards/about-project-boards) feature, but even with that, there's a ton of different ways everything could be set up.

Anything more specific you had in mind? I'm down with pretty much anything, especially in the proposal stage of everything, I'm just a bit curious on what exactly you'd want.

**Comment by Aeres-u99** [comment]
Seems like the GH project feature is an excellent choice for the same. Specifically speaking, we need a bit of organization with things, Issues can be categorized as Features | Bugs | support | documentation and regarding we can have a page where in things which are in pending state will be mentioned, once completed the person who completed it (or group of people who did it) will be mentioned near the task. 
For Bugs we can have a similar outline on GH project page and that will help us in tracking it down, I think there should be a bot which would facillitate sending the bugs to the GH page so that new contributors or us don't have to shuffle through issues pages searching for bugs to fix. 
For support guideline, My plan is to have a matrix channel/group with the bridge to telegram group, this would enable us to help the very novice user to directly communicate with Users and ask for help. This is basically more like a marketing strategy that few projects employ to get famous among the groups. 
For documentation, we have a lot of things that need attention, We will simply list out the things that we feel are missing out in the documentation and whenever we are free from bugs and issues we will look into them. The more important thing is placement of Feature request, we will have to make sure that we are not just implementing features with bugs, instead going with fixing the existing features and implementing new ones. 
I feel we should have a guideline on what should become the part of the project and what should not, like clearly addition of a new flag to facillitate generation of signature (just an example) would be great, but having a feature like parsing the apt indexes and install deb is unnecessary as they are already being provided by other packages (apt and dpkg respectively) (this is just an example) Other than these a guideline on coding, like 4 space indentation, global variables in capitals and things like that would allow us to have consistency across the code that we write. 

I will probably start looking into issues as of now, and give these a thorough thought and let me know in case there is anything thats unnecessary in my suggestion, and lets fix that and go ahead with the roadmap thing

**Comment by hwittenborn** [comment]
> Issues can be categorized as Features | Bugs | support | documentation

This part is already done with the tags and issue templates that are set up.

---

> once completed the person who completed it (or group of people who did it) will be mentioned near the task.

i.e. you'd like just people to be credited for their contributions? If so, I don't think it'd need anything more than possibly comments about contributors in source files, and maybe a master `CONTRIBUTORS` file listing all contributors or anything similar.

---

> For Bugs we can have a similar outline on GH project page and that will help us in tracking it down, I think there should be a bot which would facillitate sending the bugs to the GH page so that new contributors or us don't have to shuffle through issues pages searching for bugs to fix.

Again, the tags and issue templates allivate this imo (just search for issues by specific tags, and you can get issues, feature requests, etc).

Knowing if bugs/feature requests are being worked on could be nice, but things like PRs that are tied to issues could also help out at bit.

---

> For support guideline, My plan is to have a matrix channel/group with the bridge to telegram group, this would enable us to help the very novice user to directly communicate with Users and ask for help.

Isn't this already in effect? `#makedeb-project:hunterwittenborn.com` is already the main support room for the project, it just also includes general discussion on everything as well.

---

>  For documentation, we have a lot of things that need attention, We will simply list out the things that we feel are missing out in the documentation and whenever we are free from bugs and issues we will look into them. The more important thing is placement of Feature request, we will have to make sure that we are not just implementing features with bugs, instead going with fixing the existing features and implementing new ones.

Anyone can make an issue/PR in [makedeb/makedeb-docs](https://github.com/makedeb/makedeb-docs) if they had any gripes with the current documentation.

If anyone were working on a "feature request" that could be solved with a bug fix, we could just tell said person that - is that part related to the documentation at all?

---

> I feel we should have a guideline on what should become the part of the project and what should not, like clearly addition of a new flag to facillitate generation of signature (just an example) would be great, but having a feature like parsing the apt indexes and install deb is unnecessary as they are already being provided by other packages (apt and dpkg respectively) (this is just an example)

With your second example anyway (parsing apt indexes and installing debs), that's already mostly assumed, right? (I'm thinking your talking about searching and installing any random package? Let me know if otherwise.)

I don't think there's much that would really be added that I wouldn't accept, as long as people literally aren't thinking (if you got any other examples though, feel free to mention them).

---

> Other than these a guideline on coding, like 4 space indentation, global variables in capitals and things like that would allow us to have consistency across the code that we write.

+1 on that, a coding style guideline would definitely help for people who are contributing. I'll probably end up not making it a priority unless a decent amount of people start contributing though.

---

Sorry for making you read that super long text, hopefully that gets somewhere with some of your questions though.

**Comment by Aeres-u99** [comment]
Ahh yes yes, I have seen that. But you see the Bug and Issue template would be created mostly from the end users, and then not every bug needs on the spot attention and not every feature needs on the spot attention (again), and in due time we would be overflowing with the issues because issues won't only cover the bugs but also features|support|documentation things, mostly this would only lead to more chaos. What I would recommend is having another page where in the top 10 of the bugs|features|issues that any of the maintainers feel needs attention should be there, so that we know how much we have progressed in each of the things (Believe me it keeps things sane in the long run). Simply put, we need a type of prioritization, so that we know what needs to be done first. 

And as for coding guidelines we should make it a priority in my opinion, mostly for the sanity reasons or before we end up getting caught up in Issues and bugs and forget about it altogether. Also, we should have an Idea of the code structure without going through the code, Something like this [EAF Hacking](https://github.com/manateelazycat/emacs-application-framework/wiki/Hacking) is really cool and it helps a lot in going quickly through things (and also in case 10 years down the line if we forget what we wrote :P )

Yes rest of the things sounds just perfect. 
And long passages are just fine, Don't mind reading

**Comment by hwittenborn** [comment]
Regarding contributing guidelines, you can now find one [here](https://github.com/makedeb/makedeb/blob/alpha/CONTRIBUTING.md), which contains some info on the directory structure and code styling. I've just started it, so if you feel like anything is missing, feel free to make any comments.

I'll go ahead and get some roadmaps done under Projects real quick.

**Comment by hwittenborn** [comment]
Instead of a roadmap actually, what if we used a label guideline such as the one [matrix-org/synapse](https://github.com/matrix-org/synapse/issues/9460) uses?

Or if that would still not work, I'm still open to be convinced.

**Comment by Aeres-u99** [comment]
Yes! a label guideline like the synapse is just the perfect thing, and regarding the contributing guideline I would like to include more details, something like how the arguments are being parsed and how they are being passed, quite recently while looking into -g I was overwhelmed by the structure, I could figure out that I am supposed to put the generate_checksum.sh in functions but and how and where it should go was still a mystery, Nevertheless I will figure out the coding structure and add certain changes in the guideline and that will fix the things. For now we are all sorted on the ideas that I had. 

Just a tl;dr for the entire discussing:

1. [Coding Guideline and Contribution with notes on hacking the code](https://github.com/makedeb/makedeb/blob/alpha/CONTRIBUTING.md)
2. [Documentation and Docs related issues](https://github.com/makedeb/makedeb-docs)
3. Matrix Synapse like structure to streamline the Issues and prioritize the features. 

Correct me if I am wrong anywhere in tl;dr and then feel free to close the issue. 
For further discussion if needed any, we can just have it in the closed issue as such. I am glad that we sorted out these things as it will remove lots of confusion around

**Comment by hwittenborn** [comment]
Probably something wrong (on my end) if you couldn't figure out where arguments are being parsed.

With the `-g` example anyway, you can find it [here](https://github.com/makedeb/makedeb/blob/alpha/src/functions/arg_check.sh#L22), [here](https://github.com/makedeb/makedeb/blob/alpha/src/functions/arg_check.sh#L44), and [here](https://github.com/makedeb/makedeb/blob/alpha/src/functions/arg_check.sh#L72) (all are in `src/functions/arg_check.sh`), the first two of which just assign a variable, and then causes the command in the third one to run when it finds that the variable has been set.

Lastly, I'll work on getting the issue labels done (and better contributing guidelines, lol).

**Comment by Aeres-u99** [comment]
Oh, I had that figured out but I couldnt figure out how I am supposed to handle the function and where to put the code :P will probably check your commit regarding that.

**Comment by hwittenborn** [comment]
For the placement of the function file (`.sh` file) itself, anywhere in `src/functions` will do.

The main script (`makedeb.sh`) automatically runs `source file.sh` on every script in `functions`, so all you need to to is create the file, and you should be good to go.

Guess I need more documentation...

**Comment by Aeres-u99** [comment]
:P yeap, and I should finish the diagram for layout idea as well. My apologies for delays on my part, I have been occupied much lately.

**Comment by hwittenborn** [comment]
There's nothing wrong on your end, you aren't taking a while on anything.

I'm a bit busy anyway too, so I guess it evens out.

**Comment by hwittenborn** [comment]
Most of this has currently been solved, so I'm going to go ahead and close the issue for now.

If there's anything in particular you're still wanting, go ahead and open up a new issue and we can go from there.
